spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/SparkTable.java

```diff
@@ -247,9 +247,6 @@ public ScanBuilder newScanBuilder(CaseInsensitiveStringMap options) {
 
   @Override
   public WriteBuilder newWriteBuilder(LogicalWriteInfo info) {
-    Preconditions.checkArgument(
-        snapshotId == null, "Cannot write to table at a specific snapshot: %s", snapshotId);
```

Comment:
Why is this no longer valid? I think that we do not want to write to a specific snapshot. Is branch somehow passed as the snapshot ID?

Reply:
After looking into this a bit more, I think this is incorrect. The `snapshotId` is set when the table is loaded using time travel syntax. I don't think that we want to allow that.

Reply:
@rdblue Can we add more checks that if the snapshot Id is the tip of the branch, then writing to branch is supported ?

Reply:
Looks like this isn't an issue. I reverted this change and ran `TestSparkDataWrite` and everything passes. Let's revert this and run CI. If there are other issues outside of that test class, I'll take a look.

Reply:
@rdblue @amogh-jahagirdar if bug fix for read by snapshot ref gets merged https://github.com/apache/iceberg/pull/6717, then write to branch snapshot will fail as per test https://github.com/apache/iceberg/blob/master/spark/v3.3/spark/src/test/java/org/apache/iceberg/spark/sql/TestDeleteFrom.java#L78. That's because of the above condition.

Reply:
I cherry picked the commit from #6717 along with the proposed change to keep the condition as is, the TestDeleteFrom tests still pass for me consistently, if you're able to see failures could you post the failure when you get a chance?

Reply:
if you cherrypicked #6717 on top of current the test TestSparkDataWrite will fail

Reply:
I see now. I think this goes back to @rdblue point here https://github.com/apache/iceberg/pull/6651#discussion_r1085898935. 

So instead of passing in the snapshot through the constructor we should just directly pass in all the options. The snapshot ID can be resolved in the scan itself. That seems cleaner and I think will get us out of the removals we do  here https://github.com/apache/iceberg/pull/6717/files#diff-d278772fd3dc1431367d81a075a79404d9e1acff28fab611ad4e3d1343133596R357. 

Does that make sense? I think we can discuss more on https://github.com/apache/iceberg/pull/6717/files# since 

1.) We should get in #6717 first since that's an important fix.
2.) In #6717 we can handle all this refactorings and then rebase this PR to unblock

Reply:
Actually since the issue is that `catalog.loadTable(table)` interprets the branch option as the branch read option couldn't we just have a different config name when doing writes?

Reply:
@namrathamyske Yeah just updated to use the name `write-branch` and tests are passing. The issue is the name 'branch' is used for both read and write options, and when the loadTable is performed when doing the write , it treats it as a time travel. we should disambiguate the two. I think we should actually call it something else for the write case. `write-branch` kinda sounds odd to me tbh, maybe we go with `toBranch`. toBranch would be consistent with what's at the API and what's being done in the [Flink PR](https://github.com/apache/iceberg/pull/6660)

Reply:
But i think we can't disregard calling loadTable wrt to ref passed. Later in future when we implement session configs for testing `INSERT` `DELETE` operations, there is lot of overlap between read and write. Spark logical plans call the `SparkScanBuilder` https://github.com/apache/iceberg/blob/32a8ef52ddf20aa2068dfff8f9e73bd5d27ef610/spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/SparkScanBuilder.java#L260, https://github.com/apache/iceberg/blob/32a8ef52ddf20aa2068dfff8f9e73bd5d27ef610/spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/SparkScanBuilder.java#L424, https://github.com/apache/iceberg/blob/32a8ef52ddf20aa2068dfff8f9e73bd5d27ef610/spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/SparkScanBuilder.java#L393. which should use the time travel config. SparkCopyOnWrite, SparkMergeOnRead have respective scanner for each of them which inherit from SparkScanBuilder. I will include the changes in this PR. Its still WIP.

Reply:
Good point @namrathamyske , I was a bit short sighted we actually do want to leverage the statistics for the specific snapshot for writes. These statistics would be used during the scan itself (for example MERGE INTO branch) . So either we 1.) seek a good way to differentiate between a time travel query where the write shouldn't be able to be applied and an intentional write on a branch or 2.) we just relax the check that a snapshot is set as you did earlier.

Reply:
@rdblue @amogh-jahagirdar @jackye1995 this is still an open item for this PR get merged. I would prefer to go with second option. But let me know otherwise!